datalad/distribution/dataset.py

```diff
@@ -187,22 +189,61 @@ def path(self):
 
-        If creating an instance of GitRepo is guaranteed to be really cheap
-        this could also serve as a test whether a repo is present.
+        If testing the validity of an instance of GitRepo is guaranteed to be
+        really cheap this could also serve as a test whether a repo is present.
+
+        Note, that this property is evaluated every time it is used. If used
+        multiple times within a function it's probably a good idea to store its
+        value in a local variable and use this variable instead.
 
         Returns
         -------
-        GitRepo
+        GitRepo or AnnexRepo
         """
 
-        # Note: lazy loading was disabled, since this is provided by the
-        # flyweight pattern already and a possible invalidation of an existing
-        # instance has to be done therein.
-        # TODO: Still this is somewhat problematic. We can't invalidate strong
-        # references
-
+        # If we already got a *Repo instance, check whether it's still valid;
+        # Note, that this basically does part of the testing that would
+        # (implicitly) be done in the loop below again. So, there's still
+        # potential to speed up when we actually need to get a new instance
+        # (or none). But it's still faster for the vast majority of cases.
+        #
+        # TODO: Dig deeper into it and melt with new instance guessing. This
+        # should also involve to reduce redundancy of testing such things from
+        # within Flyweight.__call__, AnnexRepo.__init__ and GitRepo.__init__!
+        #
+        # Also note, that this could be forged into a single big condition, but
+        # that is hard to read and we should be well aware of the actual
+        # criteria here:
+        if self._repo is not None and realpath(self.path) == self._repo.path:
+            # we got a repo and path references still match
+            if isinstance(self._repo, AnnexRepo):
+                # it's supposed to be an annex
+                if self._repo is AnnexRepo._unique_instances.get(
+                        self._repo.path, None) and \
+                        AnnexRepo.is_valid_repo(self._repo.path,
+                                                allow_noninitialized=True):
+                    # it's still the object registered as flyweight and it's a
+                    # valid annex repo
+                    return self._repo
+            elif isinstance(self._repo, GitRepo):
+                # it's supposed to be a plain git
+                if self._repo is GitRepo._unique_instances.get(
+                        self._repo.path, None) and \
+                        GitRepo.is_valid_repo(self._repo.path) and not \
+                        self._repo.is_with_annex():
+                    # it's still the object registered as flyweight, it's a
+                    # valid git repo and it hasn't turned into an annex
+                    return self._repo
```

Comment:
in all of the above, if we did not care about auto Git->Annex upgrade (which is irrelevant if it is already an annex), wouldn't it be sufficient to store in Dataset  `._repo.inode` when assigning new `._repo` (e.g. as `._repo_inode`) and comparing it for a given path upon invocation?  should probably be faster than all of the checks above...?

Reply:
> in all of the above, if we did not care about auto Git->Annex upgrade

We did. Line 235: `self._repo.is_with_annex()`.

Re inode: That might be a way to clarify some things faster, yes. But still not all of it. If you uninstall a subdataset, the directory the Repo instance is pointing to, remains. But it's not a valid repo anymore. And: Remember - I want to do more about the performance of it. Just not in this PR.

Reply:
I am not following this close enough to make a meaningful contribution to the discussion, but `inode` sounds like it won't work on Windows. While I am working on avoiding GitPy code in -revolution already, introducing it this deep in the machine would make it substantially harder.